DataDog/datadog-agent
# [system-probe] Add outgoing subnet info for network flows

### What does this PR do?

This PR adds outgoing subnet info for connections to the system-probe. Each connection object now has a `Via` field that, at least initially, will only have one field called `Subnet`, which for now only holds the `id` or `name` of the outgoing subnet for the connection.

In order to get to the outgoing subnet for a connection, we need to get the routing decision for a given `(source, destination)` tuple. This can be done either with netlink or ebpf. This PR includes both implementations, with the ebpf implementation currently commented out or not being used for now, since it requires run-time compilation support.

After retrieving the routing decision, it is cached in an LRU cache, key'ed by the (source, destination, net ns) tuple for the connection. 

From the routing decision (which gives us an interface index and gateway), we can lookup the subnet from the ec2 metadata endpoint, given the mac address of the interface. This lookup is also cached.

https://github.com/DataDog/agent-payload/pull/82 for agent-payload changes.

### Motivation

Adding gateway resolution support

### Describe your test plan

1. These instructions must be run on an ec2 AWS instance
2. Ensure system-probe is running
3. Try to connect to something that is not local, e.g. google.com via `curl google.com`
4. Get connections from system-probe: `curl -s --unix-socket /opt/datadog-agent/run/sysprobe.sock http://unix/connections?client_id=1`
5. For the connection to google.com, you should see the `routeIdx` field be non-zero. There should also be a `routes` field in the output. This will list all the subnets that are referenced in the output; the `routeIdx` value is an index into this list. Note the value of the subnet (something like `subnet-`). Also note the raddr IP address of the google.com connection.
6. run `ip route get <ip from above>`. output should be something like: 
```
root@ip-172-29-37-10:/# ip route get 172.217.15.78
172.217.15.78 via 172.29.0.1 dev eth0 src 172.29.37.10 uid 0
    cache
```
The `dev eth0` is the interface the connection went out on in the example above. This could be different on your test setup.
7. Run ``curl http://169.254.169.254/latest/meta-data/network/interfaces/macs/`cat /sys/class/net/eth0/address`/subnet-id``, substituting `eth0` for whatever the device name is on your setup. The output should match the subnet you noted down in step (5.).
